### PR TITLE
Refactor: split out flavor assigment logic

### PR DESCRIPTION
**Pull request by alculquicondor**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

Split the logic for flavor assignment into its own package.
This decoupling can lead to better unit testing and more maintainable extensions in each of the packages.

I also added an "assignment mode" into the flavor assignment, which opens the possibility to assign flavors that require preemption.

#### Which issue(s) this PR fixes:

Preparation for #83 

#### Special notes for your reviewer:

The integration tests didn't require any changes :D

### CODE REVIEW

File: pkg/scheduler/flavorassigner/flavorassigner_test.go [hunk 1]
```diff
@@ -0,0 +1,849 @@
...
+	"k8s.io/apimachinery/pkg/util/sets"
+
+	kueue "sigs.k8s.io/kueue/apis/kueue/v1alpha2"
+	"sigs.k8s.io/kueue/pkg/cache"
+	"sigs.k8s.io/kueue/pkg/util/pointer"
+	utiltesting "sigs.k8s.io/kueue/pkg/util/testing"
+	"sigs.k8s.io/kueue/pkg/workload"
+)
+
+func TestAssignFlavors(t *testing.T) {
```
Comment (ahg-g) [inline pkg/scheduler/flavorassigner/flavorassigner_test.go:38]: are the tests in this file mostly a copy?

File: pkg/scheduler/flavorassigner/flavorassigner.go [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+	// CohortFit means that there are enough unused resources in the cohort to
+	// assign this flavor.
+	CohortFit FlavorAssignmentMode = iota
+	// CohortReclaim means that there is enough unused min quota in the
+	// ClusterQueue, but it needs to be reclaimed from the cohort.
+	CohortReclaim
+	// ClusterQueuePreempt means that there is not enough min quota in the
+	// ClusterQueue, so we need to preempt other workloads in the ClusterQueue
+	// with lower priority.
+	ClusterQueuePreempt
```
[the remark this answers is not in the document]
Comment (alculquicondor) [reply pkg/scheduler/flavorassigner/flavorassigner.go]: Just left CohortFit for now.

**Pull request merged**